src/realm/exec/realm_dump.c

```diff
@@ -75,69 +86,174 @@ void dump(FILE* fp, int64_t offset, size_t sz)
     }
 }
 
-size_t dump_header(FILE* fp, int64_t offset)
+int get_header(NodeHeader* node_header, FILE* fp, int64_t offset)
 {
+    memset(node_header, 0, sizeof(NodeHeader));
     unsigned char header[8];
     do_seek(fp, offset, SEEK_SET);
-    size_t actual = fread(header, 1, 8, fp);
-    if (strncmp(header, "AAAA", 4) != 0) {
+    fread(header, 1, 8, fp);
+    if (strncmp((const char*)header, "AAAA", 4) != 0) {
         printf("Ref '0x%zx' does not point to an array\n", offset);
         dump(fp, offset, 64);
         return 0;
     }
     /* dump_buffer(header, offset, 8); */
 
-    size_t size = (header[5] << 16) + (header[6] << 8) + header[7];
+    node_header->size = (header[5] << 16) + (header[6] << 8) + header[7];
 
     unsigned flags = header[4];
-    unsigned wtype = (flags & 0x18) >> 3;
-    unsigned width = (1 << (flags & 0x07)) >> 1;
-    int is_inner = (flags & 0x80) ? 1 : 0;
-    int has_refs = (flags & 0x40) ? 1 : 0;
-    int context = (flags & 0x20) ? 1 : 0;
-
-    size_t num_bytes = 0;
-    const char* type = "";
-    switch (wtype) {
+    node_header->wtype = (flags & 0x18) >> 3;
+    node_header->width = (1 << (flags & 0x07)) >> 1;
+    node_header->is_inner = (flags & 0x80) ? 1 : 0;
```

Comment:
Also, all these masks they require some expert kind of knowledge to read them and understand them. Probably we should put all of these code inside some helper functions.